natss/docs/metrics.md

```diff
@@ -0,0 +1,93 @@
+      - source_labels: [__meta_kubernetes_namespace, __meta_kubernetes_pod_label_messaging_knative_dev_role, __meta_kubernetes_pod_label_messaging_knative_dev_channel, __meta_kubernetes_pod_container_port_name]
+        action: keep
+        regex: knative-eventing;controller;natss-channel;metrics
+      # Rename metadata labels to be reader friendly
+      - source_labels: [__meta_kubernetes_namespace]
+        target_label: namespace
+      - source_labels: [__meta_kubernetes_pod_name]
+        target_label: pod
+      - source_labels: [__meta_kubernetes_service_name]
+        target_label: service
+
+  # natsschannel_dispatcher
+  - job_name: natsschannel_dispatcher
+    scrape_interval: 3s
+    scrape_timeout: 3s
+    kubernetes_sd_configs:
+      - role: pod
+    relabel_configs:
+      # Scrape only the the targets matching the following metadata
+      - source_labels: [__meta_kubernetes_namespace,__meta_kubernetes_pod_label_messaging_knative_dev_role, __meta_kubernetes_pod_label_messaging_knative_dev_channel, __meta_kubernetes_pod_container_port_name]
+        action: keep
+        regex: knative-eventing;dispatcher;natss-channel;metrics
+      # Rename metadata labels to be reader friendly
+      - source_labels: [__meta_kubernetes_namespace]
+        target_label: namespace
+      - source_labels: [__meta_kubernetes_pod_name]
+        target_label: pod
+      - source_labels: [__meta_kubernetes_service_name]
+        target_label: service
+  ```
+
+## Restart Prometheus pods to pick up this new config map changes
+  ```bash
+  kubectl delete pods -n knative-monitoring prometheus-system-0 prometheus-system-1
+  ```
+
+## Check Prometheus jobs
+
+- Do a port-forward to access Prometheus targets endpoint:
+  ```bash
+  kubectl port-forward -n knative-monitoring \
+    $(kubectl get pods -n knative-monitoring \
+    --selector=app=prometheus --output=jsonpath="{.items[0].metadata.name}") \
+    9090
+  ```
+
+- Access Prometheus targets endpoint http://localhost:9090/targets. Both the **natsschannel_controller** and **natsschannel_dispatcher** jobs should be up.
+
+
+## Check NatssChannel controller metrics
+
+- Do a port-forward to access NatssChannel controller metrics endpoint:
+  ```bash
+  kubectl port-forward -n knative-eventing \
+    $(kubectl get pods -n knative-eventing \
+    --selector='messaging.knative.dev/channel=natss-channel,messaging.knative.dev/role=controller' -o=jsonpath='{.items[0].metadata.name}') \
+    9091:9090
+  ```
+
+- Create a test NatssChannel:
+  ```bash
+  cat << EOF | kubectl apply -f -
+  apiVersion: messaging.knative.dev/v1alpha1
+  kind: NatssChannel
+  metadata:
+    name: test-natsschannel
+    namespace: default
+  EOF
+  ```
+
+- Access NatssChannel controller metrics http://localhost:9091/metrics. The metrics mentioned here https://github.com/knative/pkg/blob/master/controller/stats_reporter.go#L35-L37 ("work_queue_depth", "reconcile_count", "reconcile_latency") should be reported.
```

Comment:
I think you are not using the controller stats_reporter from pkg...
The controller uses the stats_reporter from natsschannel, this one https://github.com/knative/eventing-contrib/blob/master/natss/pkg/reconciler/stats_reporter.go
So you are reporting natsschannel_ready_count and natsschannel_ready_latency...
Can you update the doc?

Reply:
We are using the controller impl in the pkg which in turn uses the stats_reporter in the pkg:
- https://github.com/marcobebway/eventing-contrib/blob/master/natss/cmd/channel_controller/main.go#L35
- https://github.com/marcobebway/eventing-contrib/blob/master/natss/cmd/channel_controller/main.go#L95

Reply:
Ohh my bad, you are right. 
But check that we also create the other `stats_reporter` here: https://github.com/knative/eventing-contrib/blob/ad4569bc59a2b1f5cb0bfb9343826567f7c44651/natss/pkg/reconciler/reconciler.go#L141
which reports ready counts and latencies as well. 

I think I'd rather leave this part of the documentation as:
`Access NatssChannel controller metrics http://localhost:9091/metrics.`
and not mention a specific line in the source code (https://github.com/knative/pkg/blob/master/controller/stats_reporter.go#L35-L37), as that might change, and the documentation will quickly become stale. 

Do you agree?

Reply:
Yes, I agreed. I removed it from the markdown file.

Also, I verified that we can see the following metrics:

[knative/pkg/controller/stats_reporter.go](https://github.com/knative/pkg/blob/master/controller/stats_reporter.go#L35-L37)
- work_queue_depth
- reconcile_count
- reconcile_latency

[knative/eventing-contrib/natss/pkg/reconciler/stats_reporter.go](https://github.com/knative/eventing-contrib/blob/master/natss/pkg/reconciler/stats_reporter.go#L34-L36)
- natsschannel_ready_count
- natsschannel_ready_latency


<details><summary>Click for more details</summary>

```bash
# HELP natsschannel_controller_natsschannel_ready_count Number of NatssChannel that became ready
# TYPE natsschannel_controller_natsschannel_ready_count counter
natsschannel_controller_natsschannel_ready_count{key="default/my-test-channel",reconciler="natss-ch-controller"} 1
natsschannel_controller_natsschannel_ready_count{key="test-natss/natss-kn2-ingress",reconciler="natss-ch-controller"} 1
natsschannel_controller_natsschannel_ready_count{key="test-natss/natss-kn2-trigger",reconciler="natss-ch-controller"} 1
# HELP natsschannel_controller_natsschannel_ready_latency Time it takes for a NatssChannel to become ready since created
# TYPE natsschannel_controller_natsschannel_ready_latency gauge
natsschannel_controller_natsschannel_ready_latency{key="default/my-test-channel",reconciler="natss-ch-controller"} 494
natsschannel_controller_natsschannel_ready_latency{key="test-natss/natss-kn2-ingress",reconciler="natss-ch-controller"} 804
natsschannel_controller_natsschannel_ready_latency{key="test-natss/natss-kn2-trigger",reconciler="natss-ch-controller"} 381
# HELP natsschannel_controller_reconcile_count Number of reconcile operations
# TYPE natsschannel_controller_reconcile_count counter
natsschannel_controller_reconcile_count{key="default/my-test-channel",reconciler="NatssChannels",success="true"} 3
natsschannel_controller_reconcile_count{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true"} 5
natsschannel_controller_reconcile_count{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true"} 3
# HELP natsschannel_controller_reconcile_latency Latency of reconcile operations
# TYPE natsschannel_controller_reconcile_latency histogram
natsschannel_controller_reconcile_latency_bucket{key="default/my-test-channel",reconciler="NatssChannels",success="true",le="10"} 2
natsschannel_controller_reconcile_latency_bucket{key="default/my-test-channel",reconciler="NatssChannels",success="true",le="100"} 3
natsschannel_controller_reconcile_latency_bucket{key="default/my-test-channel",reconciler="NatssChannels",success="true",le="1000"} 3
natsschannel_controller_reconcile_latency_bucket{key="default/my-test-channel",reconciler="NatssChannels",success="true",le="10000"} 3
natsschannel_controller_reconcile_latency_bucket{key="default/my-test-channel",reconciler="NatssChannels",success="true",le="30000"} 3
natsschannel_controller_reconcile_latency_bucket{key="default/my-test-channel",reconciler="NatssChannels",success="true",le="60000"} 3
natsschannel_controller_reconcile_latency_bucket{key="default/my-test-channel",reconciler="NatssChannels",success="true",le="+Inf"} 3
natsschannel_controller_reconcile_latency_sum{key="default/my-test-channel",reconciler="NatssChannels",success="true"} 33
natsschannel_controller_reconcile_latency_count{key="default/my-test-channel",reconciler="NatssChannels",success="true"} 3
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true",le="10"} 4
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true",le="100"} 5
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true",le="1000"} 5
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true",le="10000"} 5
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true",le="30000"} 5
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true",le="60000"} 5
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true",le="+Inf"} 5
natsschannel_controller_reconcile_latency_sum{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true"} 49.99999999999999
natsschannel_controller_reconcile_latency_count{key="test-natss/natss-kn2-ingress",reconciler="NatssChannels",success="true"} 5
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true",le="10"} 2
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true",le="100"} 3
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true",le="1000"} 3
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true",le="10000"} 3
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true",le="30000"} 3
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true",le="60000"} 3
natsschannel_controller_reconcile_latency_bucket{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true",le="+Inf"} 3
natsschannel_controller_reconcile_latency_sum{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true"} 27.999999999999996
natsschannel_controller_reconcile_latency_count{key="test-natss/natss-kn2-trigger",reconciler="NatssChannels",success="true"} 3
# HELP natsschannel_controller_work_queue_depth Depth of the work queue
# TYPE natsschannel_controller_work_queue_depth gauge
natsschannel_controller_work_queue_depth{reconciler="NatssChannels"} 0
```

</details>